File: backend/src/routes/queues.ts
import { Router } from 'express';
import { z } from 'zod';
import { prisma } from '../lib/prisma';
import { requireAuth } from '../middleware/auth';
import { AppError } from '../middleware/error';

const router = Router();

const slugify = (s: string) =>
  s.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '').slice(0, 60);

router.get('/', requireAuth, async (req, res) => {
  const projectId = req.query.projectId as string | undefined;
  const where = projectId ? { projectId } : {};
  const queues = await prisma.queue.findMany({
    where,
    include: { project: { select: { id: true, name: true, color: true } } },
    orderBy: { createdAt: 'desc' },
  });
  // add live stats
  const stats = await Promise.all(
    queues.map(async (q) => {
      const [pending, queued, running, completed, failed, dead] = await Promise.all([
        prisma.job.count({ where: { queueId: q.id, status: 'pending' } }),
        prisma.job.count({ where: { queueId: q.id, status: 'queued' } }),
        prisma.job.count({ where: { queueId: q.id, status: 'running' } }),
        prisma.job.count({ where: { queueId: q.id, status: 'completed' } }),
        prisma.job.count({ where: { queueId: q.id, status: 'failed' } }),
        prisma.job.count({ where: { queueId: q.id, status: 'dead' } }),
      ]);
      return { ...q, stats: { pending, queued, running, completed, failed, dead } };
    }),
  );
  res.json({ queues: stats });
});

const createSchema = z.object({
  projectId: z.string().uuid(),
  name: z.string().min(1).max(120),
  description: z.string().max(500).optional(),
  concurrency: z.number().int().min(1).max(100).optional(),
  priority: z.number().int().optional(),
  maxAttempts: z.number().int().min(1).max(20).optional(),
  backoffType: z.enum(['fixed', 'exponential']).optional(),
  backoffDelayMs: z.number().int().min(100).optional(),
});

router.post('/', requireAuth, async (req, res, next) => {
  try {
    const body = createSchema.parse(req.body);
    const slug = slugify(body.name) + '-' + Date.now().toString(36).slice(-4);
    const queue = await prisma.queue.create({
      data: {
        projectId: body.projectId,
        name: body.name,
        slug,
        description: body.description,
        concurrency: body.concurrency ?? 5,
        priority: body.priority ?? 0,
        maxAttempts: body.maxAttempts ?? 3,
        backoffType: body.backoffType ?? 'exponential',
        backoffDelayMs: body.backoffDelayMs ?? 2000,
      },
    });
    res.status(201).json({ queue });
  } catch (err) {
    next(err);
  }
});

const patchSchema = createSchema.partial().omit({ projectId: true }).extend({
  isPaused: z.boolean().optional(),
});

router.patch('/:id', requireAuth, async (req, res, next) => {
  try {
    const body = patchSchema.parse(req.body);
    const queue = await prisma.queue.update({ where: { id: req.params.id }, data: body });
    res.json({ queue });
  } catch (err) {
    next(err);
  }
});

router.post('/:id/pause', requireAuth, async (req, res) => {
  const queue = await prisma.queue.update({ where: { id: req.params.id }, data: { isPaused: true } });
  res.json({ queue });
});

router.post('/:id/resume', requireAuth, async (req, res) => {
  const queue = await prisma.queue.update({ where: { id: req.params.id }, data: { isPaused: false } });
  res.json({ queue });
});

router.delete('/:id', requireAuth, async (req, res, next) => {
  try {
    await prisma.queue.delete({ where: { id: req.params.id } });
    res.json({ ok: true });
  } catch (err) {
    next(err);
  }
});

router.get('/:id', requireAuth, async (req, res, next) => {
  try {
    const queue = await prisma.queue.findUnique({
      where: { id: req.params.id },
      include: { project: true, scheduledJobs: true },
    });
    if (!queue) throw new AppError(404, 'NOT_FOUND', 'Queue not found');
    res.json({ queue });
  } catch (err) {
    next(err);
  }
});

export default router;

File: backend_py/routes/queues.py
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional
from ..lib.db import get_db
from ..lib.deps import get_current_user
from ..lib.schemas import QueueCreatePayload, QueuePatchPayload
from ..lib.utils import slugify, now

router = APIRouter()

@router.get('/')
async def list_queues(projectId: Optional[str] = Query(None), current_user: dict = Depends(get_current_user)):
    db = get_db()
    filter_query = {'project_id': projectId} if projectId else {}
    queues = await db.queues.find(filter_query).sort('created_at', -1).to_list(None)
    enriched = []
    for queue in queues:
        pending = await db.jobs.count_documents({'queue_id': queue['id'], 'status': 'pending'})
        queued = await db.jobs.count_documents({'queue_id': queue['id'], 'status': 'queued'})
        running = await db.jobs.count_documents({'queue_id': queue['id'], 'status': 'running'})
        completed = await db.jobs.count_documents({'queue_id': queue['id'], 'status': 'completed'})
        failed = await db.jobs.count_documents({'queue_id': queue['id'], 'status': 'failed'})
        dead = await db.jobs.count_documents({'queue_id': queue['id'], 'status': 'dead'})
        project = await db.projects.find_one({'id': queue['project_id']})
        enriched.append({
            **queue,
            'project': project,
            'stats': {
                'pending': pending,
                'queued': queued,
                'running': running,
                'completed': completed,
                'failed': failed,
                'dead': dead,
            },
        })
    return {'queues': enriched}

@router.post('/')
async def create_queue(payload: QueueCreatePayload, current_user: dict = Depends(get_current_user)):
    db = get_db()
    slug = f"{slugify(payload.name)}-{now().timestamp():.0f}"[:60]
    queue = {
        'id': f'queue_{now().timestamp():.0f}_{payload.name[:6]}',
        'project_id': payload.projectId,
        'name': payload.name,
        'slug': slug,
        'description': payload.description,
        'concurrency': payload.concurrency or 5,
        'priority': payload.priority or 0,
        'max_attempts': payload.maxAttempts or 3,
        'backoff_type': payload.backoffType or 'exponential',
        'backoff_delay_ms': payload.backoffDelayMs or 2000,
        'is_paused': False,
        'created_at': now(),
        'updated_at': now(),
    }
    await db.queues.insert_one(queue)
    return {'queue': queue}

@router.patch('/{queue_id}')
async def update_queue(queue_id: str, payload: QueuePatchPayload, current_user: dict = Depends(get_current_user)):
    db = get_db()
    update = {k: v for k, v in payload.model_dump(exclude_unset=True).items() if k != 'projectId'}
    if not update:
        raise HTTPException(status_code=400, detail='No update payload')
    update['updated_at'] = now()
    result = await db.queues.update_one({'id': queue_id}, {'$set': update})
    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail='Queue not found')
    queue = await db.queues.find_one({'id': queue_id})
    return {'queue': queue}

@router.post('/{queue_id}/pause')
async def pause_queue(queue_id: str, current_user: dict = Depends(get_current_user)):
    db = get_db()
    await db.queues.update_one({'id': queue_id}, {'$set': {'is_paused': True, 'updated_at': now()}})
    queue = await db.queues.find_one({'id': queue_id})
    return {'queue': queue}

@router.post('/{queue_id}/resume')
async def resume_queue(queue_id: str, current_user: dict = Depends(get_current_user)):
    db = get_db()
    await db.queues.update_one({'id': queue_id}, {'$set': {'is_paused': False, 'updated_at': now()}})
    queue = await db.queues.find_one({'id': queue_id})
    return {'queue': queue}

@router.delete('/{queue_id}')
async def delete_queue(queue_id: str, current_user: dict = Depends(get_current_user)):
    db = get_db()
    await db.queues.delete_one({'id': queue_id})
    return {'ok': True}

@router.get('/{queue_id}')
async def get_queue(queue_id: str, current_user: dict = Depends(get_current_user)):
    db = get_db()
    queue = await db.queues.find_one({'id': queue_id})
    if not queue:
        raise HTTPException(status_code=404, detail='Queue not found')
    queue['project'] = await db.projects.find_one({'id': queue['project_id']})
    queue['scheduled_jobs'] = await db.scheduled_jobs.find({'queue_id': queue_id}).to_list(None)
    return {'queue': queue}
